### PR TITLE
feat: Snap Deals

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

File: documentation/default-curio-config.toml
Comment (LexLuthr) [inline documentation/default-curio-config.toml:1]: This file should not exist. I think you might need to rebase the PR on latest main.

File: deps/config/types.go [hunk 1]
```diff
@@ -193,6 +193,13 @@ type CurioSubsystemsConfig struct {
 	// uses all available network (or disk) bandwidth on the machine without causing bottlenecks.
 	MoveStorageMaxTasks int
 
+	EnableUpdateEncode bool
+	EnableUpdateProve  bool
+	EnableUpdateSubmit bool
+
+	UpdateEncodeMaxTasks int
+	UpdateProveMaxTasks  int
```
Comment (LexLuthr) [inline deps/config/types.go:201]: We should add some comments for these config params.

File: lib/tarutil/systar.go
Comment (LexLuthr) [inline lib/tarutil/systar.go:1]: Can we document the usage for these functions? We are expanding code base considerably and undocumented functions add more time to understand functionality and purpose of libraries.

File: market/deal_ingest_snap.go [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+					Address: market.Address,
+					Payload: payload,
+				},
+			},
+		}
+
+		piece.DealProposal = nil
+		piece.DealID = 0
+		piece.PublishCid = nil
+	}
```
Comment (LexLuthr) [inline market/deal_ingest_snap.go:237]: We should do this conversion just before the ProveUpdate message. We do the same thing for ProveCommit 3. This approach looses important metadata about the deal. Post DDO, metadata is very tricky on chain and deleting this metadata locally is not something I would recommend.

**Review by LexLuthr — CHANGES_REQUESTED** [review]
I did the first pass and major change I would recommend is proposal to Manifest conversion. I think we should do that in the last step and keep the metadata before that. Some important metadata like dealID and if deal is verified are hard to find otherwise.

File: lib/dealdata/dealdata.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+				})
+				pieceReaders = append(pieceReaders, nullreader.NewNullReader(pad.Unpadded()))
+			}
+
+			pieceInfos = append(pieceInfos, abi.PieceInfo{
+				Size:     abi.PaddedPieceSize(p.PieceSize),
+				PieceCID: c,
+			})
+
+			offset += abi.UnpaddedPieceSize(*p.DataRawSize)
```
Comment (LexLuthr) [inline lib/dealdata/dealdata.go:125]: We should fix this to use unpadded size.

File: tasks/snap/task_prove.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	}
+}
+
+func (p *ProveTask) schedule(ctx context.Context, taskFunc harmonytask.AddTaskFunc) error {
+	var tasks []struct {
+		SpID         int64 `db:"sp_id"`
+		SectorNumber int64 `db:"sector_number"`
+	}
+
+	err := p.db.Select(ctx, &tasks, `SELECT sp_id, sector_number FROM sectors_snap_pipeline WHERE after_encode = TRUE AND after_prove = FALSE AND task_id_prove IS NULL`)
```
Comment (magik6k) [inline tasks/snap/task_prove.go:143]: Note to self: This (and in other tasks) select should be inside the taskFunc transaction

File: web/api/webrpc/deals.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	CreatedAt    time.Time `db:"created_at"`
+	SnapDeals    bool      `db:"is_snap"`
+
+	PieceSizeStr string `db:"-"`
+	CreatedAtStr string `db:"-"`
+}
+
+func (a *WebRPC) DealsPending(ctx context.Context) ([]OpenDealInfo, error) {
+	var deals []OpenDealInfo
+	err := a.deps.DB.Select(ctx, &deals, `SELECT sp_id, sector_number, piece_cid, piece_size, created_at, is_snap FROM open_sector_pieces ORDER BY created_at DESC`)
```
Comment (snadrus) [inline web/api/webrpc/deals.go:29]: what makes this a snap? How do these already have their sector numbers? If that is already figured-out, why must the user select SealNow() as that should have happened by trigger already?

**Review by snadrus — CHANGES_REQUESTED** [review]

**Comment by snadrus** [comment]
Can you describe, maybe in a godoc:
- how I get pieces into being Snapped vs Sealed?
- What selects the grouping?
- What triggers begin the snapping?

File: lib/fastparamfetch/paramfetch.go [hunk 6]
```diff
@@ -307,7 +307,7 @@ func fetchWithAria2c(ctx context.Context, out, url string) error {
 		return xerrors.New("aria2c not found in PATH")
 	}
 
-	cmd := exec.CommandContext(ctx, aria2cPath, "--continue", "-x16", "-s16", "-dir", filepath.Dir(out), "-o", filepath.Base(out), url)
+	cmd := exec.CommandContext(ctx, aria2cPath, "--continue", "-x16", "-s16", "--dir", filepath.Dir(out), "-o", filepath.Base(out), url)
```
[the remark this answers is not in the document]
Comment (magik6k) [reply lib/fastparamfetch/paramfetch.go]: On main now - https://github.com/filecoin-project/curio/blob/main/lib/fastparamfetch/paramfetch.go#L310

File: lib/ffi/snap_funcs.go [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+		return cid.Cid{}, cid.Cid{}, xerrors.Errorf("mkdir update cache: %w", err)
+	}
+
+	keyPath := filepath.Join(paths.UpdateCache, "cu-sector-key.dat")           // can this be a named pipe - no, mmap in proofs
+	keyCachePath := filepath.Join(paths.UpdateCache, "cu-sector-key-fincache") // some temp copy (finalized cache directory)
+	stagedDataPath := paths.Unsealed
+
+	var cleanupStagedFiles func() error
+	{
+		// hack until we do snap encode ourselves and just call into proofs for CommR
```
[the remark this answers is not in the document]
Comment (magik6k) [reply lib/ffi/snap_funcs.go:72]: https://github.com/filecoin-project/curio/issues/92, adding as a comment

File: tasks/snap/task_encode.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	id := ids[0]
+	return &id, nil
+}
+
+func (e *EncodeTask) TypeDetails() harmonytask.TaskTypeDetails {
+	ssize := abi.SectorSize(32 << 30) // todo task details needs taskID to get correct sector size
+	if seal.IsDevnet {
+		ssize = abi.SectorSize(2 << 20)
+	}
+	gpu := 1.0
```
[the remark this answers is not in the document]
Comment (magik6k) [reply tasks/snap/task_encode.go:121]: Yes, encode builds TreeR on the update-replica

File: tasks/snap/task_movestorage.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+		return false, xerrors.Errorf("moving storage: %w", err)
+	}
+
+	_, err = m.db.Exec(ctx, `UPDATE sectors_snap_pipeline SET after_move_storage = TRUE, task_id_move_storage = NULL WHERE task_id_move_storage = $1`, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("updating task: %w", err)
+	}
+
+	return true, nil
+}
```
[the remark this answers is not in the document]
Comment (magik6k) [reply tasks/snap/task_movestorage.go:77]: GC will do that. We don't leave anything on the machine which does the encode, and the original replica is needed for at least another 900 epochs after the upgrade message lands (winningPoSt has a 900 epoch lookback, forcing us to keep the original CC replica around for a while)

File: lib/tarutil/systar.go
Comment (magik6k) [reply lib/tarutil/systar.go:1]: Added a comment

File: lib/ffi/sdr_funcs.go [hunk 10]
```diff
@@ -64,7 +64,7 @@ type storageProvider struct {
 	storageReservations *xsync.MapOf[harmonytask.TaskID, *StorageReservation]
 }
 
-func (l *storageProvider) AcquireSector(ctx context.Context, taskID *harmonytask.TaskID, sector storiface.SectorRef, existing, allocate storiface.SectorFileType, sealing storiface.PathType) (fspaths, ids storiface.SectorPaths, release func(), err error) {
+func (l *storageProvider) AcquireSector(ctx context.Context, taskID *harmonytask.TaskID, sector storiface.SectorRef, existing, allocate storiface.SectorFileType, sealing storiface.PathType) (fspaths, ids storiface.SectorPaths, release func(dontDeclace ...storiface.SectorFileType), err error) {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply lib/ffi/sdr_funcs.go]: Typo, fixed

File: market/deal_ingest_snap.go (hunk 2 above)
Comment (magik6k) [reply market/deal_ingest_snap.go:237]: Not so clear, I suspect that things will move more and more into the direction of DDO / custom markets, I don't think we need to treat the builtin markets in a glorified way

In the UI we can still decode the PAM and display the DealID

File: web/api/webrpc/deals.go (hunk 5 above)
Comment (magik6k) [reply web/api/webrpc/deals.go:29]: This isn't snap specific, applies to both snap and porep deal ingest.

File: market/seal_now.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type SealNowNodeApi interface {
+	StateMinerInfo(ctx context.Context, addr address.Address, tsk types.TipSetKey) (api.MinerInfo, error)
+	StateNetworkVersion(ctx context.Context, tsk types.TipSetKey) (networkVersion network.Version, err error)
+}
+
+func SealNow(ctx context.Context, node SealNowNodeApi, db *harmonydb.DB, act address.Address, sector abi.SectorNumber, synthetic bool) error {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply market/seal_now.go:25]: This one is a user trigger, added a comment.

**Comment by magik6k** [comment]
> how I get pieces into being Snapped vs Sealed?

This was missing, now is decided by the config

> What selects the grouping?
> What triggers begin the snapping?

Those two are the same as normal deal ingest

**Review by LexLuthr — APPROVED** [review]

File: lib/ffi/snap_funcs.go [hunk 12]
```diff
@@ -0,0 +1,324 @@
+package ffi
```
[the remark this answers is not in the document]
Comment (magik6k) [reply lib/ffi/snap_funcs.go:1]: This package is used in a ton of places, if we want to rename that should be a separate PR

**Comment by magik6k** [comment]
Tested on my setup, appears to work well.

File: tasks/snap/task_encode.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+			Storage: e.sc.Storage(e.taskToSector, storiface.FTUpdate|storiface.FTUpdateCache|storiface.FTUnsealed, storiface.FTNone, ssize, storiface.PathSealing, 1.0),
+		},
+		MaxFailures: 3,
+		IAmBored: passcall.Every(MinSnapSchedInterval, func(taskFunc harmonytask.AddTaskFunc) error {
+			return e.schedule(context.Background(), taskFunc)
+		}),
+	}
+}
+
+func (e *EncodeTask) Adder(taskFunc harmonytask.AddTaskFunc) {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply tasks/snap/task_encode.go:143]: Adds are done in IAmBored

File: tasks/seal/task_sdr.go [hunk 14]
```diff
@@ -27,14 +27,14 @@ import (
 	"github.com/filecoin-project/lotus/storage/sealer/storiface"
 )
 
-var isDevnet = lbuild.BlockDelaySecs < 30
+var IsDevnet = lbuild.BlockDelaySecs < 30
```
[the remark this answers is not in the document]
Comment (magik6k) [reply tasks/seal/task_sdr.go]: Comment on old commit? Right now the code depends on curio/build

File: market/deal_ingest_snap.go [hunk 15]
```diff
@@ -0,0 +1,572 @@
...
+			if sec.currentSize+psize <= abi.PaddedPieceSize(p.sectorSize) {
+				if vd.isVerified {
+					sectorLifeTime := sec.latestEndEpoch - sec.earliestStartEpoch
+					// Allocation's TMin must fit in sector and TMax should be at least sector lifetime or more
+					// Based on https://github.com/filecoin-project/builtin-actors/blob/a0e34d22665ac8c84f02fea8a099216f29ffaeeb/actors/verifreg/src/lib.rs#L1071-L1086
+					if sectorLifeTime <= vd.tmin && sectorLifeTime >= vd.tmax {
+						continue
+					}
+
+					//  ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////  TODO ADD SNAP SECTOR EXP CHECKS
```
[the remark this answers is not in the document]
Comment (magik6k) [reply market/deal_ingest_snap.go:378]: https://github.com/filecoin-project/curio/issues/104

File: market/deal_ingest_snap.go [hunk 16]
```diff
@@ -0,0 +1,572 @@
...
+	}
+
+	upt, err := p.getUpgradeProofType()
+	if err != nil {
+		return xerrors.Errorf("getting seal proof type: %w", err)
+	}
+
+	shouldSeal := func(sector *openSector) bool {
+		// Start sealing a sector if
+		// 1. If sector is full
```
[the remark this answers is not in the document]
Comment (magik6k) [reply market/deal_ingest_snap.go:104]: iiuc docs for snap will be written soon

**Pull request merged**